src/cmd/builtin/overlay.c

```diff
@@ -919,6 +938,136 @@ static int subcmd_disconnect (optparse_t *p, int ac, char *av[])
     return 0;
 }
 
+static struct hostlist *idset2hostlist (flux_t *h, struct idset *idset)
+{
+    struct hostlist *hl;
+    unsigned int i;
+
+    if (!(hl = hostlist_create ()))
+        log_err_exit ("cannot create hostlist");
+
+    i = idset_first (idset);
+    while (i != IDSET_INVALID_ID) {
+        const char *host = flux_get_hostbyrank (h, i);
+        if (hostlist_append (hl, host) <= 0)
+            log_msg_exit ("failed to append host %s", host);
+        i = idset_next (idset, i);
+    }
+    /* could do uniq, but we are electing not to, since multiple
+     * brokers can exist on one host.
+     */
+    // hostlist_uniq (hl);
+    hostlist_sort (hl);
+    return hl;
+}
+
+static int subcmd_whatsup (optparse_t *p, int ac, char *av[])
+{
+    flux_t *h = builtin_get_flux_handle (p);
+    bool up = optparse_hasopt (p, "up");
+    bool down = optparse_hasopt (p, "down");
+    bool ranks = optparse_hasopt (p, "ranks");
+    bool count = optparse_hasopt (p, "count");
+    flux_future_t *f;
+    const char *members, *val;
+    struct idset *idset_up, *idset_down;
+    struct hostlist *hl_up = NULL, *hl_down = NULL;
+    char *up_str = NULL, *down_str = NULL;
+    size_t up_count, down_count;
+    size_t overlay_size;
+    char *endptr;
+
+    /* get currently "up" ranks */
+    if (!(f = flux_rpc_pack (h,
+                             "groups.get",
+                             0, /* must send to rank 0 */
+                             0,
+                             "{s:s}",
+                             "name", "broker.online")))
+        log_err_exit ("groups.get");
+    if (flux_rpc_get_unpack (f, "{s:s}", "members", &members) < 0)
+        log_err_exit ("groups.get: %s", future_strerror (f, errno));
+
+    /* get max ranks size */
+    if (!(val = flux_attr_get (h, "size")))
+        log_err_exit ("cannot get size attr");
+
+    errno = 0;
+    overlay_size = strtoul (val, &endptr, 10);
+    if (errno != 0 || *endptr != '\0' || overlay_size == 0)
+        log_msg_exit ("invalid overlay size retrieved: %s", val);
+
+    if (!(idset_up = idset_decode (members)))
+        log_msg_exit ("cannot decode up idset %s", members);
+
+    /* determine down nodes */
+    if (!(idset_down = idset_create (overlay_size, 0)))
+        log_err_exit ("cannot create idset");
+    if (idset_range_set (idset_down, 0, overlay_size - 1) < 0)
+        log_err_exit ("cannot set idset range");
+    if (idset_subtract (idset_down, idset_up) < 0)
+        log_err_exit ("cannot construct idset down");
+
+    up_count = idset_count (idset_up);
+    down_count = idset_count (idset_down);
+
+    if (ranks) {
+        up_str = idset_encode (idset_up, IDSET_FLAG_RANGE);
+        down_str = idset_encode (idset_down, IDSET_FLAG_RANGE);
+    }
+    else {
+        hl_up = idset2hostlist (h, idset_up);
+        hl_down = idset2hostlist (h, idset_down);
+        up_str = hostlist_encode (hl_up);
+        down_str = hostlist_encode (hl_down);
```

Comment:
Use `flux_hostmap_lookup(3)` (you'll have to create `up_str` and `down_str` for the hostlist case, but then you can remove the local `idset2hostlist()` function which mostly duplicates the functionality of `flux_hostmap_lookup(3)`.